File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/groupsv2/DecryptedGroupChangeActionsBuilderChangeSetModifier.java
package org.whispersystems.signalservice.api.groupsv2;

import com.google.protobuf.ByteString;

import org.signal.storageservice.protos.groups.Member;
import org.signal.storageservice.protos.groups.local.DecryptedGroupChange;
import org.signal.storageservice.protos.groups.local.DecryptedMember;
import org.signal.storageservice.protos.groups.local.DecryptedRequestingMember;

import java.util.ArrayList;
import java.util.List;

final class DecryptedGroupChangeActionsBuilderChangeSetModifier implements ChangeSetModifier {
  private final DecryptedGroupChange.Builder result;

  public DecryptedGroupChangeActionsBuilderChangeSetModifier(DecryptedGroupChange.Builder result) {
    this.result = result;
  }

  @Override
  public void removeAddMembers(int i) {
    result.removeNewMembers(i);
  }

  @Override
  public void moveAddToPromote(int i) {
    DecryptedMember addMemberAction = result.getNewMembersList().get(i);
    result.removeNewMembers(i);
    result.addPromotePendingMembers(addMemberAction);
  }

  @Override
  public void removeDeleteMembers(int i) {
    List<ByteString> newList = removeIndexFromByteStringList(result.getDeleteMembersList(), i);

    result.clearDeleteMembers()
          .addAllDeleteMembers(newList);
  }

  @Override
  public void removeModifyMemberRoles(int i) {
    result.removeModifyMemberRoles(i);
  }

  @Override
  public void removeModifyMemberProfileKeys(int i) {
    result.removeModifiedProfileKeys(i);
  }

  @Override
  public void removeAddPendingMembers(int i) {
    result.removeNewPendingMembers(i);
  }

  @Override
  public void removeDeletePendingMembers(int i) {
    result.removeDeletePendingMembers(i);
  }

  @Override
  public void removePromotePendingMembers(int i) {
    result.removePromotePendingMembers(i);
  }

  @Override
  public void clearModifyTitle() {
    result.clearNewTitle();
  }

  @Override
  public void clearModifyAvatar() {
    result.clearNewAvatar();
  }

  @Override
  public void clearModifyDisappearingMessagesTimer() {
    result.clearNewTimer();
  }

  @Override
  public void clearModifyAttributesAccess() {
    result.clearNewAttributeAccess();
  }

  @Override
  public void clearModifyMemberAccess() {
    result.clearNewMemberAccess();
  }

  @Override
  public void clearModifyAddFromInviteLinkAccess() {
    result.clearNewInviteLinkAccess();
  }

  @Override
  public void removeAddRequestingMembers(int i) {
    result.removeNewRequestingMembers(i);
  }

  @Override
  public void moveAddRequestingMembersToPromote(int i) {
    DecryptedRequestingMember addMemberAction = result.getNewRequestingMembersList().get(i);
    result.removeNewRequestingMembers(i);

    DecryptedMember build = DecryptedMember.newBuilder()
                                           .setUuid(addMemberAction.getUuid())
                                           .setProfileKey(addMemberAction.getProfileKey())
                                           .setRole(Member.Role.DEFAULT).build();

    result.addPromotePendingMembers(0, build);
  }

  @Override
  public void removeDeleteRequestingMembers(int i) {
    List<ByteString> newList = removeIndexFromByteStringList(result.getDeleteRequestingMembersList(), i);

    result.clearDeleteRequestingMembers()
          .addAllDeleteRequestingMembers(newList);
  }

  @Override
  public void removePromoteRequestingMembers(int i) {
    result.removePromoteRequestingMembers(i);
  }

  private static List<ByteString> removeIndexFromByteStringList(List<ByteString> byteStrings, int i) {
    List<ByteString> modifiedList = new ArrayList<>(byteStrings);

    modifiedList.remove(i);

    return modifiedList;
  }
}

File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/groupsv2/DecryptedGroupChangeActionsBuilderChangeSetModifier.kt
package org.whispersystems.signalservice.api.groupsv2

import org.signal.storageservice.protos.groups.AccessControl
import org.signal.storageservice.protos.groups.Member
import org.signal.storageservice.protos.groups.local.DecryptedGroupChange
import org.signal.storageservice.protos.groups.local.DecryptedMember
import org.signal.storageservice.protos.groups.local.DecryptedRequestingMember
import org.signal.storageservice.protos.groups.local.EnabledState

internal class DecryptedGroupChangeActionsBuilderChangeSetModifier(private val result: DecryptedGroupChange.Builder) : ChangeSetModifier {
  override fun removeAddMembers(i: Int) {
    result.newMembers = result.newMembers.removeIndex(i)
  }

  override fun moveAddToPromote(i: Int) {
    val addMemberAction: DecryptedMember = result.newMembers[i]
    result.newMembers = result.newMembers.removeIndex(i)
    result.promotePendingMembers += addMemberAction
  }

  override fun removeDeleteMembers(i: Int) {
    result.deleteMembers = result.deleteMembers.removeIndex(i)
  }

  override fun removeModifyMemberRoles(i: Int) {
    result.modifyMemberRoles = result.modifyMemberRoles.removeIndex(i)
  }

  override fun removeModifyMemberProfileKeys(i: Int) {
    result.modifiedProfileKeys = result.modifiedProfileKeys.removeIndex(i)
  }

  override fun removeAddPendingMembers(i: Int) {
    result.newPendingMembers = result.newPendingMembers.removeIndex(i)
  }

  override fun removeDeletePendingMembers(i: Int) {
    result.deletePendingMembers = result.deletePendingMembers.removeIndex(i)
  }

  override fun removePromotePendingMembers(i: Int) {
    result.promotePendingMembers = result.promotePendingMembers.removeIndex(i)
  }

  override fun clearModifyTitle() {
    result.newTitle = null
  }

  override fun clearModifyAvatar() {
    result.newAvatar = null
  }

  override fun clearModifyDisappearingMessagesTimer() {
    result.newTimer = null
  }

  override fun clearModifyAttributesAccess() {
    result.newAttributeAccess = AccessControl.AccessRequired.UNKNOWN
  }

  override fun clearModifyMemberAccess() {
    result.newMemberAccess = AccessControl.AccessRequired.UNKNOWN
  }

  override fun clearModifyAddFromInviteLinkAccess() {
    result.newInviteLinkAccess = AccessControl.AccessRequired.UNKNOWN
  }

  override fun removeAddRequestingMembers(i: Int) {
    result.newRequestingMembers = result.newRequestingMembers.removeIndex(i)
  }

  override fun moveAddRequestingMembersToPromote(i: Int) {
    val addMemberAction: DecryptedRequestingMember = result.newRequestingMembers[i]
    result.newRequestingMembers = result.newRequestingMembers.removeIndex(i)

    val promote = DecryptedMember(
      aciBytes = addMemberAction.aciBytes,
      profileKey = addMemberAction.profileKey,
      role = Member.Role.DEFAULT
    )
    result.promotePendingMembers += promote
  }

  override fun removeDeleteRequestingMembers(i: Int) {
    result.deleteRequestingMembers = result.deleteRequestingMembers.removeIndex(i)
  }

  override fun removePromoteRequestingMembers(i: Int) {
    result.promoteRequestingMembers = result.promoteRequestingMembers.removeIndex(i)
  }

  override fun clearModifyDescription() {
    result.newDescription = null
  }

  override fun clearModifyAnnouncementsOnly() {
    result.newIsAnnouncementGroup = EnabledState.UNKNOWN
  }

  override fun removeAddBannedMembers(i: Int) {
    result.newBannedMembers = result.newBannedMembers.removeIndex(i)
  }

  override fun removeDeleteBannedMembers(i: Int) {
    result.deleteBannedMembers = result.deleteBannedMembers.removeIndex(i)
  }

  override fun removePromotePendingPniAciMembers(i: Int) {
    result.promotePendingPniAciMembers = result.promotePendingPniAciMembers.removeIndex(i)
  }

  private fun <T> List<T>.removeIndex(i: Int): List<T> {
    val modifiedList = this.toMutableList()
    modifiedList.removeAt(i)
    return modifiedList
  }
}
